crates/flox-rust-sdk/src/models/environment/core_environment.rs

```diff
@@ -0,0 +1,361 @@
+    }
+
+    /// Get the path to the lockfile
+    ///
+    /// Note: may not exist
+    fn lockfile_path(&self) -> PathBuf {
+        self.env_dir.join(LOCKFILE_FILENAME)
+    }
+
+    /// Read the manifest file
+    fn manifest_content(&self) -> Result<String, EnvironmentError2> {
+        fs::read_to_string(self.manifest_path()).map_err(EnvironmentError2::ReadManifest)
+    }
+
+    /// Lock the environment.
+    ///
+    /// This updates the lock if it exists, or generates a new one if it doesn't.
+    ///
+    /// Technically this does write to disk as a side effect for now.
+    /// It's included in the [ReadOnly] struct for ergonomic reasons
+    /// and because it doesn't modify the manifest.
+    ///
+    /// todo: should we always write the lockfile to disk?
+    pub fn lock(&mut self, flox: &Flox) -> Result<LockedManifest, EnvironmentError2> {
+        let manifest_path = self.manifest_path();
+        let lockfile_path = self.lockfile_path();
+        let maybe_lockfile = if lockfile_path.exists() {
+            debug!("found existing lockfile: {}", lockfile_path.display());
+            Some(lockfile_path.as_ref())
+        } else {
+            debug!("no existing lockfile found");
+            None
+        };
+        let lockfile = LockedManifest::lock_manifest(
+            Path::new(&*PKGDB_BIN),
+            &manifest_path,
+            maybe_lockfile,
+            &global_manifest_path(flox),
+        )?;
+
+        // Write the lockfile to disk
+        // todo: do we always want to do this?
+        debug!("generated lockfile, writing to {}", lockfile_path.display());
+        std::fs::write(&lockfile_path, lockfile.to_string())
+            .map_err(EnvironmentError2::WriteLockfile)?;
+
+        Ok(lockfile)
+    }
+
+    /// Build the environment, [Self::lock] if necessary.
+    ///
+    /// Technically this does write to disk as a side effect for now.
+    /// It's included in the [ReadOnly] struct for ergonomic reasons
+    /// and because it doesn't modify the manifest.
+    ///
+    /// Does not link the environment to an out path.
+    /// Linking should be done explicitly by the caller using [Self::link].
+    ///
+    /// todo: should we always write the lockfile to disk?
+    pub fn build(&mut self, flox: &Flox) -> Result<PathBuf, EnvironmentError2> {
+        let lockfile = self.lock(flox)?;
+
+        debug!(
+            "building environment: system={}, lockfilePath={}",
+            &flox.system,
+            self.lockfile_path().display()
+        );
+
+        let store_path = lockfile.build(Path::new(&*ENV_BUILDER_BIN), None)?;
+
+        debug!(
+            "built locked environment, store path={}",
+            store_path.display()
+        );
+
+        Ok(store_path)
+    }
+
+    /// Create a new out-link for the environment at the given path.
+    ///
+    /// Builds the environment if necessary.
+    /// todo: should we always build implicitly?
+    pub fn link(
```

Comment:
nonblocking: not even sure if it would be better, but would it be better to just pass `build()` a path to link to? Also not sure if it would be simpler/no need to change now, but if build returned a store path, we could `nix-store --add-root` instead of building again